code/modules/mob/living/carbon/xenomorph/castes/wraith/abilities_wraith.dm

```diff
@@ -0,0 +1,460 @@
+	add_cooldown()
+
+
+/datum/action/xeno_action/phase_shift/proc/phase_shift_deactivate()
+
+	var/mob/living/carbon/xenomorph/wraith/TI = owner
+
+	if(!TI) //Sanity
+		return
+
+	TI.status_flags = initial(TI.status_flags) //Become merely mortal again
+	TI.resistance_flags = initial(TI.resistance_flags)
+	TI.density = initial(TI.density)
+	TI.throwpass = initial(TI.throwpass)
+	TI.alpha = initial(TI.alpha) //Become opaque
+
+	playsound(owner, "sound/effects/escape_pod_launch.ogg", 25, 0, 1)
+
+	var/current_turf = get_turf(TI)
+
+	if(isclosedturf(current_turf) || isspaceturf(current_turf)) //So we rematerialized in a solid wall/space for some reason; Darwin award winner folks.
+		to_chat(TI, "<span class='highdanger'>As we idiotically rematerialize in an obviously unsafe position, we revert to where we slipped out of reality through sheer instinct, albeit at great cost.</span>")
+		TI.adjustFireLoss(TI.maxHealth * 0.5, TRUE) //Lose half of our maximum health
+		TI.apply_status_effect(STATUS_EFFECT_STUN, 10 SECONDS) //That oughta teach them.
+		TI.forceMove(starting_turf)
+		teleport_debuff_aoe(TI) //Debuff when we reappear
+		starting_turf = null
+		return
+
+	var/distance = get_dist(current_turf, starting_turf)
+	var/phase_shift_stun_time = clamp(0.5 SECONDS, distance * 0.1 SECONDS, 3 SECONDS) //Recovery time
+	TI.apply_status_effect(STATUS_EFFECT_STUN, phase_shift_stun_time * 2)
+
+	TI.visible_message("<span class='warning'>[TI.name] form wavers and becomes opaque.</span>", \
+	"<span class='highdanger'>We phase back into reality, our mind reeling from the experience. We estimate we will take [phase_shift_stun_time * 0.1] seconds to recover!</span>")
+
+	starting_turf = null
+
+/datum/action/xeno_action/evasion/on_cooldown_finish()
+
+	if(!owner) //Sanity
+		return
+
+	to_chat(owner, "<span class='xenonotice'>We are able to fade from reality again.</span>")
+	owner.playsound_local(owner, 'sound/effects/xeno_newlarva.ogg', 25, 0, 1)
+	return ..()
+
+
+// ***************************************
+// *********** Blink
+// ***************************************
+/datum/action/xeno_action/activable/blink
+	name = "Blink"
+	action_icon_state = "blink"
+	mechanics_text = "We teleport ourselves a short distance to a location within line of sight."
+	use_state_flags = XABB_TURF_TARGET
+	plasma_cost = 50
+	cooldown_timer = 0.5 SECONDS
+	keybind_signal = COMSIG_XENOABILITY_BLINK
+
+/datum/action/xeno_action/activable/blink/can_use_ability(atom/A, silent = FALSE, override_flags)
+	. = ..()
+	var/turf/T = get_turf(A)
+
+	if(!T)
+		to_chat(owner, "<span class='xenowarning'>Even we cannot go there!</span>")
+		return FALSE
+
+	if(isclosedturf(T) || isspaceturf(T))
+		to_chat(owner, "<span class='xenowarning'>We cannot blink here!</span>")
+		return FALSE
+
+	if(!owner.line_of_sight(T) || A.opacity) //Needs to be in line of sight.
+		to_chat(owner, "<span class='xenowarning'>We can't blink without line of sight to our destination!</span>")
+		return FALSE
```

Comment:
Why A

Reply:
So we can't teleport into vision blockers